### DISCUSSION TITLE
Should `x-model` with `select` have the ability to decide if it should use `.text` or `.value`

### DISCUSSION THREAD

**Discussion by byjpr** [alpinejs/alpine#2790] [3. Feature ideas]
I'm wondering if I should write a PR to create a modifier that would allow the selection of either the value of the selected option or text.

Currently, the code will check for value, and if no value, it will take the text. 

https://github.com/alpinejs/alpine/blob/5cb40460d744c2a840f21d937e71e202471f7e21/packages/alpinejs/src/directives/x-model.js#L102-L104

### Current behaviour
If you have the following code:

```html
<select name="user[role]" x-model="select">
  <option value="role1" selected="">Admin</option>
  <option value="role2">User</option>
</select>
```

The text inside `<span x-text="select"></span>` would be `role1`.

However, 

```html
<select name="user[role]" x-model="select">
  <option selected="">Admin</option>
  <option>User</option>
</select>
```

The text inside `<span x-text="select"></span>` would be `Admin`.

### What would be introduced?
Another modifier that would set preference for `text`. E.G: `x-model.text="select"`

**Comment by SimoTod** [comment]
I think the code you attached is only relevant for multi-select.
In general, it's not an Alpine thing, the value of a standard select box is always the text if a value attribute is not provided: https://codepen.io/SimoTod/pen/zYpzwgP

**Reply by byjpr** [reply]
You're correct. I thought the check for multi was an `or` statement. 🤦🏼

Absolutely. With Alpine `x-model` it will prefer the value attribute, and then fall back to the inner text if the value attribute is not set. This is to do with the way you described, the browser will use the inner text when the option tag doesn't have a value attribute. I'm describing having an override. Pushed up an example last night in #2792.

**Reply by asbjornu** [reply]
I came here wanting to get out both the `text` and `value` of the selected `option` only to find I have to choose between the two. If I manually write JavaScript to find the selected `<option>`, I can extract its `value` as well as its `text`, but with Alpine, this isn't possible?

**Reply by ekwoka** [reply]
As it stands, Alpine is handling it in the same way it is natively handled.

**Reply by asbjornu** [reply]
@ekwoka, what does that mean? Natively, in plain HTML and JavaScript, I can do this:

```html
<form>
    <select id="numbers">
        <option value="1">One</option>
        <option value="2">Two</option>
        <option value="3">Three</option>
    </select>
</form>
<script>
    document.getElementById('numbers').addEventListener('change', function(event) {
        console.log(event.target.value, event.target.selectedOptions[0].text);
    });
</script>
```

And `1 One`, `2 Two`, `3 Three` will be logged to the console depending on which `<option>` I select. How can I achieve the same using Alpine?

**Reply by ekwoka** [reply]
Yes, because you're actively accessing both form the OPTION element.

The behavior of the SELECT element is that its value will be the values if present otherwise the text.

The select element itself will not provide you the value and text content of the options.

That's the native behavior that Alpine replicates.

**Reply by asbjornu** [reply]
Ok, that makes sense. But don't you agree that it's a valid request to have Alpine expose both, somehow? Having to dive down into plain JS to dig out data from the DOM is very clunky compared to just using Alpine's fabulous built-in `x-model` and `x-bind` attributes.

**Reply by SimoTod** [reply]
@asbjornu you can do exactly the same. It's not any clunkier than the original example
```
<form x-data="{value: null, text: null}">
    <select id="numbers" x-model="value" @change="text = $event.target.selectedOptions[0].text">
        <option value="1">One</option>
        <option value="2">Two</option>
        <option value="3">Three</option>
    </select>
    <p>Value: <span x-text="value"><span></p>
    <p>Text: <span x-text="text"><span></p>
</form>
```

**Reply by asbjornu** [reply]
@SimoTod, I did not know about `$event` and did not think about using both `x-model` and `@change` simultaneously. 👍🏼 A bit clunky still, but only compared to how clean most Alpine code is. Thanks! 🙏🏼

**Reply by ekwoka** [reply]
It can be reasonable to want both.

I think the focus of the library (as any should) is staying fairly "spec compliant" where behaviors are defined.

That's the primary.

Then other behaviors would need to evaluate if there is a need for something more, how common that need is, and how painful it is to implement in user land.
